components/encryption/src/manager/mod.rs

```diff
@@ -198,24 +235,28 @@ impl Dicts {
         Ok(())
     }
 
-    fn link_file(&mut self, src_fname: &str, dst_fname: &str) -> Result<()> {
-        let file = match self.file_dict.files.get(src_fname) {
+    fn link_file(&self, src_fname: &str, dst_fname: &str) -> Result<()> {
+        let _lock = self.file_lock.lock().unwrap();
+        let mut file_dict = self.file_dict.lock().unwrap();
+        let file = match file_dict.files.get(src_fname) {
             Some(file_info) => file_info.clone(),
             None => {
                 // Could be a plaintext file not tracked by file dictionary.
                 info!("link untracked plaintext file"; "src" => src_fname, "dst" => dst_fname);
                 return Ok(());
             }
         };
-        if self.file_dict.files.get(dst_fname).is_some() {
+        if file_dict.files.get(dst_fname).is_some() {
             return Err(Error::Io(IoError::new(
                 ErrorKind::AlreadyExists,
                 format!("file already exists, {}", dst_fname),
             )));
         }
         let method = file.method;
-        self.file_dict.files.insert(dst_fname.to_owned(), file);
+        file_dict.files.insert(dst_fname.to_owned(), file);
+        drop(file_dict);
```

Comment:
remove the use of `drop` for better readability - with scoping it is easier to know the scope of the lock.